src/js/toggle.js

```diff
@@ -0,0 +1,123 @@
+import Target from './target';
+
+class Toggle {
+
+	constructor(toggleEl, config) {
+		if (!Toggle._targets) {
+			Toggle._targets = new Map();
+		}
+
+		if (!toggleEl) {
+			return;
+		} else if (!(toggleEl instanceof HTMLElement)) {
+			toggleEl = document.querySelector(toggleEl);
+		}
+
+		if (toggleEl.hasAttribute('data-o-toggle--js')) {
+			return;
+		}
+
+		if (!config) {
+			config = {};
+			// Try to get config set declaratively on the element
+			Array.prototype.forEach.call(toggleEl.attributes, (attr) => {
+				if (attr.name.indexOf('data-o-toggle') === 0) {
+					// Remove the prefix part of the data attribute name
+					const key = attr.name.replace('data-o-toggle-', '');
+					try {
+						// If it's a JSON, a boolean or a number, we want it stored like that, and not as a string
+						// We also replace all ' with " so JSON strings are parsed correctly
+						config[key] = JSON.parse(attr.value.replace(/\'/g, '"'));
+					} catch (e) {
+						config[key] = attr.value;
+					}
+				}
+			});
+		}
+
+		this.callback = config.callback;
+		if (typeof this.callback === 'string') {
+			this.callback = new Function(this.callback);
+		}
+
+		this.toggleEl = toggleEl;
+
+		if (this.toggleEl.nodeName === 'A') {
+			this.toggleEl.setAttribute('role', 'button');
+		}
+
+		this.toggle = this.toggle.bind(this);
+		this.toggleEl.addEventListener('click', this.toggle);
+
+		this.toggleEl.setAttribute('data-o-toggle--js', 'true');
+
+		this.targetEl = config.target;
+		if (!(this.targetEl instanceof HTMLElement)) {
+			this.targetEl = document.querySelector(this.targetEl);
+		}
+
+		if (Toggle._targets.get(this.targetEl) === undefined) {
+			this.target = new Toggle.Target(this);
+			Toggle._targets.set(this.targetEl, this.target);
+		} else {
+			this.target = Toggle._targets.get(this.targetEl);
+		}
+
+		this.target.addToggle(this);
+		this.target.close();
+	}
+
+	open() {
+		this.toggleEl.setAttribute('aria-expanded', 'true');
+	}
+
+	close() {
+		this.toggleEl.setAttribute('aria-expanded', 'false');
+	}
+
+	toggle(e) {
+
+		this.target.toggle();
+
+		if(e) {
+			e.preventDefault();
+		}
+
+		if (this.callback){
```

Comment:
Toggle.toggle is bound in the constructor at https://github.com/Financial-Times/o-toggle/pull/11/files#diff-b0c558eea0ef24ca97b2408369732c8eR49 so this isn't a bug.

I'll add a comment to Toggle.toggle() and you're right that we're not actually testing this behaviour, so I'll add a test which binds Toggle.toggle  to an event and then trigger that event to check the context is correct.